client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java

```diff
@@ -118,6 +127,74 @@ static Request index(IndexRequest indexRequest) {
+        parameters.withTimeout(updateRequest.timeout());
+        parameters.withRefreshPolicy(updateRequest.getRefreshPolicy());
+        parameters.withWaitForActiveShards(updateRequest.waitForActiveShards());
+        parameters.withDocAsUpsert(updateRequest.docAsUpsert());
+        parameters.withFetchSourceContext(updateRequest.fetchSource());
+        parameters.withRetryOnConflict(updateRequest.retryOnConflict());
+        parameters.withVersion(updateRequest.version());
+        parameters.withVersionType(updateRequest.versionType());
+
+        XContentType xContentType = null;
+        if (updateRequest.doc() != null) {
+            xContentType = updateRequest.doc().getContentType();
+        } else if (updateRequest.upsertRequest() != null) {
+            xContentType = updateRequest.upsertRequest().getContentType();
+        } else {
+            xContentType = Requests.INDEX_CONTENT_TYPE;
+        }
+
+        return new Request(HttpPost.METHOD_NAME, endpoint, parameters.getParams(), toHttpEntity(updateRequest, xContentType));
+    }
+
+    static ByteArrayEntity toHttpEntity(UpdateRequest updateRequest, XContentType xContentType) {
+        try (XContentBuilder builder = XContentBuilder.builder(xContentType.xContent())) {
+            builder.startObject();
+            if (updateRequest.docAsUpsert()) {
+                builder.field("doc_as_upsert", updateRequest.docAsUpsert());
+            }
+            IndexRequest doc = updateRequest.doc();
+            if (doc != null) {
+                try (XContentParser parser = createParser(NamedXContentRegistry.EMPTY, doc.source(), doc.getContentType())) {
+                    builder.field("doc").copyCurrentStructure(parser);
+                }
+            }
+            Script script = updateRequest.script();
+            if (script != null) {
+                builder.field("script", script);
+            }
+            IndexRequest upsert = updateRequest.upsertRequest();
+            if (upsert != null) {
+                try (XContentParser parser = createParser(NamedXContentRegistry.EMPTY, upsert.source(), upsert.getContentType())) {
+                    builder.field("upsert").copyCurrentStructure(parser);
+                }
+            }
+            if (updateRequest.scriptedUpsert()) {
+                builder.field("scripted_upsert", updateRequest.scriptedUpsert());
+            }
+            if (updateRequest.detectNoop() == false) {
+                builder.field("detect_noop", updateRequest.detectNoop());
+            }
+            if (updateRequest.fetchSource() != null) {
+                builder.field("_source", updateRequest.fetchSource());
+            }
+            builder.endObject();
+
+            BytesRef requestBody = builder.bytes().toBytesRef();
+            ContentType contentType = ContentType.create(xContentType.mediaType());
+            return new ByteArrayEntity(requestBody.bytes, requestBody.offset, requestBody.length, contentType);
```

Comment:
should UpdateRequest implement `ToXContent` instead? we will find this in other requests too (e.g. bulk), we should decide where to put this code for all those.